src/rust/src/tracing_ffi.rs

```diff
@@ -0,0 +1,563 @@
+        reload_handle,
+    }))
+}
+
+#[no_mangle]
+pub extern "C" fn tracing_free(handle: *mut TracingHandle) {
+    drop(unsafe { Box::from_raw(handle) });
+}
+
+#[no_mangle]
+pub extern "C" fn tracing_reload(handle: *mut TracingHandle, new_filter: *const c_char) {
+    let handle = unsafe { &mut *handle };
+    let new_filter = unsafe { CStr::from_ptr(new_filter) }
+        .to_str()
+        .expect("new filter should be a valid string");
+
+    let new_filter = EnvFilter::new(new_filter);
+    handle.reload_handle.reload(new_filter).unwrap();
+}
+
+pub struct FfiCallsite {
+    interest: AtomicUsize,
+    meta: Option<Metadata<'static>>,
+    registration: Once,
+    fields: Vec<&'static str>,
+}
+
+impl FfiCallsite {
+    fn new(fields: Vec<&'static str>) -> Self {
+        FfiCallsite {
+            interest: AtomicUsize::new(0),
+            meta: None,
+            registration: Once::new(),
+            fields,
+        }
+    }
+
+    #[inline(always)]
+    fn is_enabled(&self) -> bool {
+        let interest = self.interest();
+        if interest.is_always() {
+            return true;
+        }
+        if interest.is_never() {
+            return false;
+        }
+
+        tracing::dispatcher::get_default(|current| current.enabled(self.metadata()))
+    }
+
+    #[inline(always)]
+    pub fn register(&'static self) {
+        self.registration
+            .call_once(|| tracing::callsite::register(self));
+    }
+
+    #[inline(always)]
+    fn interest(&self) -> Interest {
+        match self.interest.load(Ordering::Relaxed) {
+            0 => Interest::never(),
+            2 => Interest::always(),
+            _ => Interest::sometimes(),
+        }
+    }
+}
+
+impl Callsite for FfiCallsite {
+    fn set_interest(&self, interest: Interest) {
+        let interest = match () {
+            _ if interest.is_never() => 0,
+            _ if interest.is_always() => 2,
+            _ => 1,
+        };
+        self.interest.store(interest, Ordering::SeqCst);
+    }
+
+    #[inline(always)]
+    fn metadata(&self) -> &Metadata<'static> {
+        self.meta.as_ref().unwrap()
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn tracing_callsite(
+    name: *const c_char,
+    target: *const c_char,
+    level: *const c_char,
+    file: *const c_char,
+    line: u32,
+    fields: *const *const c_char,
+    fields_len: usize,
+    is_span: bool,
+) -> *mut FfiCallsite {
+    let name = unsafe { CStr::from_ptr(name) }.to_str().unwrap();
+    let target = unsafe { CStr::from_ptr(target) }.to_str().unwrap();
+    let level = unsafe { CStr::from_ptr(level) }.to_str().unwrap();
+    let file = unsafe { CStr::from_ptr(file) }.to_str().unwrap();
+
+    let fields = unsafe { slice::from_raw_parts(fields, fields_len) };
+    let fields = fields
+        .iter()
+        .map(|&p| unsafe { CStr::from_ptr(p) })
+        .map(|cs| cs.to_str())
+        .collect::<Result<Vec<_>, _>>()
+        .unwrap();
+
+    let level = level.parse().unwrap();
```

Comment:
Naively, it seems like you _should_ be able to just expose the `Level` type in `tracing` over FFI as a C enum. The inner type [has the `repr(usize)` attribute](https://github.com/tokio-rs/tracing/blob/0da8e3f37636845ece6bdceeecb8a0946cc77076/tracing-core/src/metadata.rs#L334-L357), so it should be a valid C enum. I think we may need to add `#[repr(transparent)]` to the [outer type](https://github.com/tokio-rs/tracing/blob/0da8e3f37636845ece6bdceeecb8a0946cc77076/tracing-core/src/metadata.rs#L96-L98) to ensure that this is FFI-safe, but I would happily accept a PR that adds that --- the `LevelFilter` type already has it.

Reply:
Alternatively, we could probably just expose a C-compatible enum in the FFI crate, and convert it to/from the `tracing::Level` type in a `match` here. 

I don't think the performance implications of using `Level::from_str` here are terribly significant; this only occurs once per callsite construction, so it's not a hot path. However, using an enum on the C++ API side would make it a _little_ more difficult to accidentally pass an invalid level.